packages/@lwc/engine-dom/src/stylesheets.ts

```diff
@@ -0,0 +1,41 @@
+import { create, isUndefined } from '@lwc/shared';
+
+const elementsToStyleContents = new WeakMap<Element, { [content: string]: true } | undefined>();
+const elementsToStyleElements = new WeakMap<Element, HTMLStyleElement | undefined>();
+
+const globalStylesheetsParentElement: Element = document.head || document.body || document;
+
+if (process.env.NODE_ENV === 'development') {
+    // @ts-ignore
+    window.__lwcResetGlobalStylesheets = () => {
+        elementsToStyleContents.delete(globalStylesheetsParentElement);
+        elementsToStyleElements.delete(globalStylesheetsParentElement);
+    };
+}
+
+export function insertStylesheet(element: Element, content: string): void {
+    let styleContents = elementsToStyleContents.get(element);
+    if (isUndefined(styleContents)) {
+        styleContents = create(null);
+        elementsToStyleContents.set(element, styleContents);
+    }
+    if (!isUndefined(styleContents![content])) {
+        return;
+    }
+    styleContents![content] = true;
+
+    let style = elementsToStyleElements.get(element);
+    if (isUndefined(style)) {
+        style = document.createElement('style');
+        style.type = 'text/css';
+        style.textContent = content;
+        element.insertBefore(style, element.firstChild);
+        elementsToStyleElements.set(element, style);
+    } else {
+        style.textContent += '\n' + content;
+    }
```

Comment:
I'm not sure about the perf of creating multiple `<style>`s versus having one `<style>` and appending to its `textContent`. It may turn out the former is faster, especially if `textContent` has to go through Locker. But my hope is that, in Chrome anyway, we can just switch to constructable stylesheets and bypass the DOM altogether.